### PR TITLE
[SYCL][LIBCLC][NATIVECPU] Add aarch64 target triple for Native CPU

### PR DESCRIPTION
**Pull request by PietroGhg**
This PR adds support to building Native CPU for `aarch64` on an `aarch64` host:
* the target triple for `libclc` is based by default on `LLVM_TARGET_TRIPLE` (this can be overridden using the `--native_cpu_libclc_targets` option in `configure.py`)
* since both `x86_64` and `aarch64` require the same specializations in libclc on linux, the `x86_64-unknown-linux` folder has been renamed `native_cpu-unknown-linux`, and the folder gets added to sources list if either `x86_64-*-linux` or `aarch64-*-linux` are added as libclc targets

### CODE REVIEW

**Comment by PietroGhg** [comment]
@intel/dpcpp-devops-reviewers the [Check code formatting](https://github.com/intel/llvm/actions/runs/9224271535/job/25379203694?pr=13911) job is requesting formatting changes to `configure.py`, I'm not sure if I should go ahead and apply them

**Comment by aelovikov-intel** [comment]
> @intel/dpcpp-devops-reviewers the [Check code formatting](https://github.com/intel/llvm/actions/runs/9224271535/job/25379203694?pr=13911) job is requesting formatting changes to `configure.py`, I'm not sure if I should go ahead and apply them

I'd say the middle (`if args.native_cpu:`) chunk only. The rest should probably go into a separate PR.

File: buildbot/configure.py [hunk 1]
```diff
@@ -276,6 +280,7 @@ def main():
     parser.add_argument("--disable-preview-lib", action='store_true', help="Disable building of the SYCL runtime major release preview library")
     parser.add_argument("--disable-fusion", action="store_true", help="Disable the kernel fusion JIT compiler")
     parser.add_argument("--add_security_flags", type=str, choices=['none', 'default', 'sanitize'], default=None, help="Enables security flags for compile & link. Two values are supported: 'default' and 'sanitize'. 'Sanitize' option is an extension of 'default' set.")
+    parser.add_argument('--native_cpu_libclc_targets', help='Target triples for libclc, used by the Native CPU backend')
```
Comment (frasercrmck) [inline buildbot/configure.py:283]: Seems like hyphens are the norm here, except for `add_security_flags`. Thus I'd suggest `native-cpu-libclc-targets`

File: libclc/CMakeLists.txt [hunk 2]
```diff
@@ -337,6 +348,10 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
         list( APPEND flags -Xclang -target-feature -Xclang +avx
             -Xclang -fsycl-is-native-cpu
             -Xclang -target-feature -Xclang +avx512f)
+      elseif( ARCH  STREQUAL aarch64)
+        list( APPEND flags -Xclang -fsycl-is-native-cpu)
+      else()
```
Comment (frasercrmck) [inline libclc/CMakeLists.txt:353]: Do we need this `else()`? We already have `set (flags )` above.

File: libclc/CMakeLists.txt [hunk 3]
```diff
@@ -370,7 +385,10 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
     elseif( ARCH STREQUAL amdgcn )
       set( opt_flags -O3 --amdgpu-oclc-reflect-enable=false )
     elseif( ARCH STREQUAL x86_64)
-      set( opt_flags )
+      set( opt_flags -O3 )
+      set( has_distinct_generic_addrspace FALSE )
+    elseif( ARCH STREQUAL "aarch64")
```
Comment (frasercrmck) [inline libclc/CMakeLists.txt:390]: Maybe leave out these quotation marks for consistency.

File: libclc/CMakeLists.txt [hunk 4]
```diff
@@ -297,6 +304,10 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
     set( DARCH ${ARCH} )
   endif()
 
+  if( (ARCH  STREQUAL x86_64 OR ARCH  STREQUAL aarch64) AND OS STREQUAL "linux")
```
Comment (frasercrmck) [inline libclc/CMakeLists.txt:307]: What would happen if you try and build this on something other than `x86_64` or `aarch64`, or don't use `linux`? Would the build fail? Would it build silently and do the wrong thing? I think we should think about how to safeguard the build from unsupported native-cpu configurations.

File: buildbot/configure.py (hunk 1 above)
Comment (frasercrmck) [reply buildbot/configure.py:283]: We should probably also rename `add_security_flags` for the users' sake.

File: libclc/CMakeLists.txt (hunk 4 above)
Comment (frasercrmck) [reply libclc/CMakeLists.txt:307]: Also there are double spaces here, and inconsistent string quoting.

File: libclc/CMakeLists.txt [hunk 5]
```diff
@@ -370,7 +385,10 @@ foreach( t ${LIBCLC_TARGETS_TO_BUILD} )
     elseif( ARCH STREQUAL amdgcn )
       set( opt_flags -O3 --amdgpu-oclc-reflect-enable=false )
     elseif( ARCH STREQUAL x86_64)
```
Comment (frasercrmck) [inline libclc/CMakeLists.txt:387]: This is the second case I see where we're still doing separate checks for `x86_64` and `aarch64` which share a lot of logic. I think we might want to think about having a common variable set for when we're compiling for native-cpu and check that at the top level.

Then, if required, we could do additional target-specific logic inside that.

``` cmake
elseif( NATIVE_CPU )
  set( opt_flags -O3 )
  set( has_distinct_generic_addrspace FALSE )
```

``` cmake
elseif( NATIVE_CPU )
  list( APPEND flags -Xclang -fsycl-is-native-cpu)
  if( ARCH STREQUAL x86_64)
    list( APPEND flags -Xclang -target-feature -Xclang +avx -Xclang -target-feature -Xclang +avx512f)
  endif()
```

... that sort of thing.

That would keep the number of changes if/when we add a third/fourth native-cpu target to a minimum.

File: buildbot/configure.py (hunk 1 above)
Comment (PietroGhg) [reply buildbot/configure.py]: I went for `native-cpu-libclc-targets`, thank you. (I'd rather not rename `add_security_flags` as part of this PR)

File: libclc/CMakeLists.txt (hunk 2 above)
Comment (PietroGhg) [reply libclc/CMakeLists.txt]: Messed up when rebasing, thanks for spotting it

File: libclc/CMakeLists.txt (hunk 4 above)
Comment (PietroGhg) [reply libclc/CMakeLists.txt]: This is a bit tricky, I've added a warning that is triggered when building libclc for a "native cpu arch" on an OS != `linux`. Other cases are a bit hard to detect: different triples may or may not be pre-existing DPC++ targets (`ptx`, `amdgcn` etc), and at this stage I think that (for Native CPU), we would like them to fall back to `generic` and see what happens

**Review by frasercrmck — APPROVED** [review]
LGTM

File: buildbot/configure.py (hunk 1 above)
Comment (frasercrmck) [reply buildbot/configure.py]: Yeah, renaming that flag isn't for this PR :+1:

**Comment by PietroGhg** [comment]
@intel/dpcpp-devops-reviewers could you take a look at this PR? The formatting job is failing as discussed [here](https://github.com/intel/llvm/pull/13911#issuecomment-2129884214), but everything else is ready for review. Thank you :)

**Comment by PietroGhg** [comment]
@intel/dpcpp-devops-reviewers I've added a commit that changes the default `host-target` from `X86` to `host`, the behaviour should be the same on x86 machines, but the "appropriate" host target should be set of different archs, let me know if it's ok

**Review by aelovikov-intel — APPROVED** [review]
`configure.py` changes LGTM. I expect other files are owned by cuda/native_cpu folks.

**Review by uwedolinsky — APPROVED** [review]

**Comment by PietroGhg** [comment]
@intel/llvm-gatekeepers this looks ready to merge, the failure in the formatting job is unrelated to the changes in the PR, see https://github.com/intel/llvm/pull/13911#issuecomment-2129884214

**Pull request merged**